src/utils.ts

```diff
@@ -898,97 +898,224 @@ export function deepCopy<T>(value: T): T {
 }
 
 /** @internal */
-const kBuffers = Symbol('buffers');
-/** @internal */
-const kLength = Symbol('length');
+type ListNode<T> = { value: T; next: ListNode<T>; prev: ListNode<T> };
```

Comment:
I think Neal's approach is correct.  With the exception of during initialization, a Node will never have `null` `next` or `prev`s.  If we adjust the `ListNode` type to account for null pointers, then we need to check for null everywhere we use them.

Reply:
This would be a fine argument if the ListNode<T> was returned to the users, but it's actually just used for the internal implementation of the list, and saying that the things that are currently typed as `ListNode<T>` always have something in them when the list methods interact with them is simply not true. The point of the TS type here is to help make sure that the internal implementation of the list accounts for all possible values of ListNode<T>. My suggestion is to explicitly define `head: HeadNode<T> | NullNode` where `HeadNode<T> = { value: null; next: ListNode<T>; prev: ListNode<T> }` and `NullNode = { value: null; next: NullNode; prev: NullNode }`. It would make it more clear that the head never contains an actual value of its own. Furthermore, setting `prev:` and `next:` of `ListNode<T>` to be `ListNode<T> | HeadNode<T>` would likewise account for those scenarios of the node in question being at the start or end of the list.

Reply:
I've added some better typing, so now that our head node is special it moves some of the casting but I think it improves the understanding here that head is special and holds null and the rest are the items in the list. 

> `NullNode = { value: null; next: NullNode; prev: NullNode }`

Trying this doesn't get type value. In theory `node !== this head` should narrow away the HeadNode to make node only a ListNode, thus making it's value the correct T, but this reference check isn't something the language supports. (Unless I'm missing something)

Reply:
From slack: some more improvements were just pushed to rid us of some casting and the casting that is left is clear where the edge case is with narrowing the type